src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -154,11 +160,31 @@ private void RequestWakeup()
             {
                 unsafe
                 {
+                    VerboseTrace("Writing to wakeup pipe");
                     byte b = 1;
                     while ((Interop.CheckIo((long)Interop.libc.write(_requestWakeupPipeFd, &b, (IntPtr)1)))) ;
                 }
             }
 
+            /// <summary>Clears data from the wakeup pipe.</summary>
+            /// <remarks>
+            /// This must only be called when we know there's data to be read.
+            /// The MultiAgent could easily deadlock if it's called when there's no data in the pipe.
+            /// </remarks>
+            private unsafe void ReadFromWakeupPipeWhenKnownToContainData()
+            {
+                // It's possible but unlikely that there will be tons of extra data in the pipe, 
+                // more than we end up reading out here (it's unlikely because we only write a byte to the pipe when 
+                // transitioning from 0 to 1 incoming request).  In that unlikely event, the worst
+                // case will be that the next one or more waits will wake up immediately, with each one
+                // subsequently clearing out more of the pipe.
+                const int ClearBufferSize = 64; // sufficiently large to clear the pipe in any normal case
+                byte* clearBuf = stackalloc byte[ClearBufferSize];
+                long bytesRead;
+                while (Interop.CheckIo(bytesRead = (long)Interop.libc.read(_wakeupRequestedPipeFd, clearBuf, (IntPtr)ClearBufferSize))) ;
+                VerboseTraceIf(bytesRead > 1, "Read more than one byte from wakeup pipe: " + bytesRead);
```

Comment:
No, the loop is just there to handle the error case of EINTR, in which case we retry the read.